### PR TITLE
rats-tls: add user callback for OPA

### PR DESCRIPTION
**Pull request by YangLiang3**
Fixes: #1297
Signed-off-by: Liang Yang <liang3.yang@intel.com>

### CODE REVIEW

**Comment by YangLiang3** [comment]
Please @liangzhou121 let's co-debug this feature firstly. If it has no any problem, then we can merge it

File: rats-tls/src/include/enclave-tls/api.h [hunk 1]
```diff
@@ -64,6 +64,45 @@ typedef struct {
...
+        int user_data_size;
+        char* user_data;
+        int unhashed_size;
+        char* unhashed;
+};
+
+typedef enum {
+        SGX_ECDSA = 1,
+        TDX
+}enclave_evidence_type_t;
```
Comment (jiazhang0) [inline rats-tls/src/include/enclave-tls/api.h:93]: Some format issues.

**Comment by jiazhang0** [comment]
The commit title doesn't need to mention OPA. This is a common mechanism for all possible upper verification handlers.

File: rats-tls/src/api/enclave_tls_init.c [hunk 2]
```diff
@@ -14,7 +14,8 @@
 #include "internal/attester.h"
 #include "internal/verifier.h"
 
-enclave_tls_err_t enclave_tls_init(const enclave_tls_conf_t *conf, enclave_tls_handle *handle)
+enclave_tls_err_t enclave_tls_init(const enclave_tls_conf_t *conf, enclave_tls_handle *handle,
+				   callback_Type cb)
```
Comment (jiazhang0) [inline rats-tls/src/api/enclave_tls_init.c:18]: callback_Type is not a regular naming scheme.

Comment (jiazhang0) [reply rats-tls/src/api/enclave_tls_init.c:18]: Why not adding a new API call enclave_tls_set_verification_callback()?

File: rats-tls/src/include/enclave-tls/api.h [hunk 3]
```diff
@@ -64,6 +64,42 @@ typedef struct {
...
 } enclave_tls_conf_t;
 
+struct etls_sgx_evidence {
+	uint8_t *mrEnclave;
+	uint8_t *mrSigner;
+	uint32_t productId;
+	uint32_t securityVersion;
+	uint8_t *attributes;
+	size_t collateral_size;
+	char *collateral;
```
Comment (jiazhang0) [inline rats-tls/src/include/enclave-tls/api.h:74]: plz fix the indent issue.

File: rats-tls/src/include/enclave-tls/api.h [hunk 4]
```diff
@@ -64,6 +64,42 @@ typedef struct {
...
+	enclave_evidence_type_t type;
+	struct ehd_s ehd;
+	int quote_size;
+	char *quote;
+	union {
+		struct etls_sgx_evidence *sgx;
+		struct etls_tdx_evidence *tdx;
+	};
+};
+
```
Comment (jiazhang0) [inline rats-tls/src/include/enclave-tls/api.h:102]: ditto

File: rats-tls/src/tls_wrappers/openssl/un_negotiate.c [hunk 5]
```diff
@@ -323,12 +313,29 @@ int verify_certificate(int preverify, X509_STORE_CTX *ctx)
...
+	if (!ev) {
+		ETLS_ERR("failed to cmalloc etls_evidence\n");
+		return 0;
+	}
 
-	if (find_oid(cert, (const char *)ecdsa_quote_oid) == SSL_SUCCESS)
+	if (find_oid(cert, (const char *)ecdsa_quote_oid) == SSL_SUCCESS) {
+		sgx_quote3_t quote3;
+
+		memset(&quote3, 0, sizeof(quote3));
```
Comment (jiazhang0) [inline rats-tls/src/tls_wrappers/openssl/un_negotiate.c:325]: why requiring this?

File: rats-tls/src/tls_wrappers/openssl/un_negotiate.c [hunk 6]
```diff
@@ -344,5 +351,9 @@ int verify_certificate(int preverify, X509_STORE_CTX *ctx)
 		return 0;
 	}
 
+	memset(&ev->ehd, 0, sizeof(ev->ehd));
```
Comment (jiazhang0) [inline rats-tls/src/tls_wrappers/openssl/un_negotiate.c:354]: why requiring this?

File: rats-tls/src/tls_wrappers/openssl/un_negotiate.c [hunk 7]
```diff
@@ -323,12 +313,29 @@ int verify_certificate(int preverify, X509_STORE_CTX *ctx)
...
+		memset(&quote3, 0, sizeof(quote3));
+		memcpy(&quote3, evidence.ecdsa.quote, sizeof(quote3));
+		ev->sgx->mrEnclave = quote3.report_body.mr_enclave.m;
+		ev->sgx->mrSigner = quote3.report_body.mr_signer.m;
+		ev->sgx->productId = quote3.report_body.isv_prod_id;
+		ev->sgx->securityVersion = quote3.report_body.isv_svn;
+		ev->sgx->attributes = (char *)&(quote3.report_body.attributes);
+		ev->type = SGX_ECDSA;
+		ev->quote = evidence.ecdsa.quote;
+		ev->quote_size = evidence.ecdsa.quote_len;
```
Comment (jiazhang0) [inline rats-tls/src/tls_wrappers/openssl/un_negotiate.c:334]: Why not use evidence.ecdsa.quote as `sgx_quote3_t *quote3`?

File: rats-tls/src/tls_wrappers/openssl/un_negotiate.c [hunk 8]
```diff
@@ -344,5 +351,9 @@ int verify_certificate(int preverify, X509_STORE_CTX *ctx)
 		return 0;
 	}
 
+	memset(&ev->ehd, 0, sizeof(ev->ehd));
+	if (!tls_ctx->etls_handle->user_callback)
+		tls_ctx->etls_handle->user_callback(ev);
```
Comment (jiazhang0) [inline rats-tls/src/tls_wrappers/openssl/un_negotiate.c:356]: ev is not freed after used.

**Review by jiazhang0 — CHANGES_REQUESTED** [review]

File: rats-tls/src/include/enclave-tls/api.h [hunk 9]
```diff
@@ -64,6 +64,42 @@ typedef struct {
...
+typedef enum { SGX_ECDSA = 1, TDX } enclave_evidence_type_t;
+
+typedef struct etls_evidence {
+	enclave_evidence_type_t type;
+	ehd_t ehd;
+	int quote_size;
+	char *quote;
+	union {
+		etls_sgx_evidence_t sgx;
```
Comment (jiazhang0) [inline rats-tls/src/include/enclave-tls/api.h:98]: rename to `etls_sgx_ecdsa_evidence_t sgx_ecdsa;`

File: rats-tls/samples/enclave-tls-server/server.c [hunk 10]
```diff
@@ -351,7 +351,7 @@ int main(int argc, char **argv)
 		}
 	} while (opt != -1);
 
-        global_log_level = log_level;
+	global_log_level = log_level;
```
Comment (jiazhang0) [inline rats-tls/samples/enclave-tls-server/server.c:354]: Could you move above changes to another commit?

File: rats-tls/src/include/enclave-tls/api.h (hunk 9 above)
Comment (YangLiang3) [reply rats-tls/src/include/enclave-tls/api.h:98]: etls_sgx_evidence_t is not just for ecdsa, it can be used later by Local attestation as scalable.

File: rats-tls/src/tls_wrappers/openssl/un_negotiate.c [hunk 11]
```diff
@@ -344,5 +334,27 @@ int verify_certificate(int preverify, X509_STORE_CTX *ctx)
 		return 0;
 	}
 
+	if (!strncmp(evidence.type, "sgx_ecdsa", sizeof(evidence.type))) {
+		etls_evidence_t ev;
+		sgx_quote3_t *quote3 = (sgx_quote3_t *)evidence.ecdsa.quote;
+
+		ev.sgx.mrEnclave = (char *)quote3->report_body.mr_enclave.m;
```
Comment (jiazhang0) [inline rats-tls/src/tls_wrappers/openssl/un_negotiate.c:341]: Please fix the naming scheme.

**Review by haosanzi — APPROVED** [review]
